### PR TITLE
PDEP-1: Purpose and guidelines for pandas enhancement proposals

### LINKED ISSUE #28568: Pandas Enhancement Proposals?
I was wondering if there has been discussion about having dedicated RFC / Proposals for larger changes to pandas? 

For example, after a rather heated discussion in #24046, @wesm [asked](https://github.com/pandas-dev/pandas/issues/24046#issuecomment-515492965):
> I question whether GitHub issues was the appropriate venue for this discussion compared with some form of RFC / design document.

I wanted to come back to this and maybe write such a design document, and therefore wanted to check what the thoughts of the core team are about this? Needless to say, the obvious candidates to model this after would be Python's [PEPs](https://www.python.org/dev/peps/) or NumPy's [NEPs](https://numpy.org/neps/), but maybe there could/should be pandas-specific departures.

I think this could benefit several high-impact design questions (e.g. around extension arrays or release plans). Actually, I have 2-3 other bigger PRs that eventually stalled in design discussion (and a few ideas of similar magnitude), where I could imagine that such a vessel would help move the discussion forward.

### PR DESCRIPTION
**Pull request by datapythonista**
Closes #28568

Initial PDEP to define purpose and guidelines for pandas enhancement proposals (equivalent to PEPs or NEPs). **Feedback very welcome**.

This PR also makes the PDEPs public in the [roadmap page](https://pandas.pydata.org/about/roadmap.html) of our website.

### CODE REVIEW

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (simonjayhawkins) [inline web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:44]:
```suggestion
an issue or creating a PDEP, it's probably safe to start by
```

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (simonjayhawkins) [inline web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:3]:
```suggestion
- Date: 21 June 2022
```

**Comment by mroeschke** [comment]
This would probably close https://github.com/pandas-dev/pandas/issues/28568

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (mroeschke) [inline web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:67]:
```suggestion
Once a PDEP is accepted, any contributions can be made toward the implementing the PDEP with an open-ended completion timeline . The
```

IMO the former wording doesn't sound that encouraging to propose PDEPs that might not be implemented

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (mroeschke) [inline web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:76]: Should the release version be included too to note which version of pandas has the complete implementation?

Comment (datapythonista) [reply web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:76]: Great idea, I included it.

**Review by mroeschke — COMMENTED** [review]
I didn't immediately see this, but it would be good to explain how an accepted/rejected decision is made. This looks to be Numpy's: https://numpy.org/neps/nep-0000.html#how-a-nep-becomes-accepted

For example, it would be good to define:

* How consensus is defined: 100%, 2/3rds, 51% approval from core devs?
* How to collect consensus: PDEP PR? Monthly dev meeting?
* Should a "call to vote" action be available to unblock perpetual discussion?

**Comment by datapythonista** [comment]
> I didn't immediately see this, but it would be good to explain how an accepted/rejected decision is made.

Very good point. I thought this was quite a big topic, and maybe more related to general governance than specific to PDEPs. That's why for now I decided to just link to the governance documentation. Based on the current governance, decisions are made by consensus, with the BDFL (Wes) having the final word when consensus is not possible.

I think the governance documentation is 7 years old. Many things changed since then, including the structure of the team (from like 3 active core devs to like 20, Python moved out of a BDFL model...). Also, besides the decision making, I think our governance seems outdated. I personally don't know most of the people in the pandas CoC committee, and I assume they are not performing any task, and probably don't even know they are part of it.

While I fully agree on what you say, I'd personally keep the discussions separate, as I think both are quiet big, and can be taken care of separately. What do you think?

**Review by mroeschke — APPROVED** [review]
Sounds good to me. Agreed that general decision making details can go under governance.

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (lithomas1) [inline web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:104]: We should probably clarify when a PDEP can be revised v.s. when a new PDEP has to be submitted superseding the old PDEP.

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (lithomas1) [inline web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:56]: Do we want to also have a public comment period on PDEP's like tensorflow has with their RFC's?

We should probably also clarify when voting happens(define what proportion of core team members need to consider a PDEP ready before voting like @mroeschke stated) and how long core team members have to vote.

**Review by lithomas1 — COMMENTED** [review]
LG overall. Just had a few questions as I was going through it.

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (datapythonista) [reply web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:53]: I'd personally leave that for later, once we need it. Those sound like good ideas to me, but I wouldn't implement them initially, I would wait to see how things work first. If we merge like PDEP per month, I think checking the last PDEP number before merging is easier than a system to autogenerate them. And for a template, I'd wait to have few actual PDEPs before deciding if it helps, or if PDEPs are too different from one to another.

Does it make sense to you to start simple and iterate later as we have more experience?

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (simonjayhawkins) [reply web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:56]: > But again, I'd start by seeing how things work, and if we see PDEPs keep open for too long, and seems like having a timeframe should help, surely worth giving it a try.

I agree, just having a PDEP is akin to a new process, so having a process for the PDEP is maybe like introducing another meta process at the same time. So in response to this comment and others related to the process, I agree with @datapythonista that to begin with we just discuss whether we want to use PDEPs, what we want out of a PDEP,  and roughly what it should contain and iterate on the gaps over time. The only caveat here, is that once someone has spent effort preparing a PDEP and it is approved that we don't have "blockers" implementing, so that does affect the approval process to some degree.

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (datapythonista) [reply web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:57]: Sounds good to me.

Regarding merging PDEPs as draft, feels like it could make discussions more difficult. Like where to have the discussion once the PR is merged. Via new PRs? Or in a second PR, not being able to comment in parts not in the diff. But I guess in some cases it can be useful, maybe a PR with different parts, and merging as a Draft once the first part has been discussed, and before working on other parts.

I'll update the document if there are no objections to using Under discussion status (or Draft if people prefer this name) until PR is Accepted. And to consider merging PRs before they are accepted when it's useful.

File: web/pandas/about/roadmap.md
Comment (rhshadrach) [inline web/pandas/about/roadmap.md:22]: I know this is bikeshedding, and certainly okay with it as-is, but maybe the abbreviation should be pdEP? My thinking here is:

 - With the "p" in pandas traditionally lowercase; as opposed to the P in Python (PEP) and N in NumPy (NEP)
 - Reflects `import pandas as pd`
 - I instinctively see "PDEP" and "P-DEP"

Comment (datapythonista) [reply web/pandas/about/roadmap.md:22]: I'd personally try to avoid adding extra complexity with the capitalization. I don't know what P-DEP is or what stands for. If it's something we want to avoid, my preference would be to use something totally different. Like PCP or PDCP (C for Change instead of E for ENHANCEMENT), or anything else. But if people prefer pdEP it, I'll update it.

**Comment by datapythonista** [comment]
I updated the PR to incorporate the feedback from comments, and I replied to the ones I didn't directly incorporate.

I also fixes couple of visualization issues, and improved a bit the styles, so the roadmap and also other pages are clearer.

Let me know of additional feedback, and feel free to approve when you're happy with getting this merged, or add a `Request changes` flagged if you see any blocker. Thanks!

**Comment by datapythonista** [comment]
@pandas-dev/pandas-core any other comment before merging this? If there are no objections I'll merge this in couple of days, and we can keep iterating as needed in separate PRs.

File: web/pandas/about/roadmap.md
Comment (jreback) [reply web/pandas/about/roadmap.md:22]: why not just PEP ; i don't think this conflicts with python enhancements and it's simpler and easier to pronounce

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (jreback) [reply web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:57]: i agree - we need a voting process here - we could emulate that of NEP which is pretty simple

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (jreback) [inline web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:57]: i am not sure this is true - sure a lot will easily be accepted but some might be controversial and ultimately not accepted

**Review by jreback — CHANGES_REQUESTED** [review]

File: web/pandas/about/roadmap.md
Comment (bashtage) [reply web/pandas/about/roadmap.md:22]: PEP could be hard to find using search engines given the ubiquity of PEP as Python Enhancement Proposal.

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (simonjayhawkins) [reply web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:57]: yeah. maybe just qualify this by changing `By default, we expect a PDEP will be accepted` to something like `By default, we expect a PDEP (with some preliminary discussion) will be accepted` 

before this text we have...

> A PDEP should be used for changes that are not
immediate and not obvious, and are expected to require a significant amount of
discussion and require detailed documentation before being implemented.

so the PEP is part of the detailed documentation and an issue should perhaps be the initial part of the discussion. (just as I would normally expect say a bug fix PR to have an associated issue)

and we also have 

> If you are unsure if you should be opening
an issue or creating a PDEP, it's probably safe to start by
[opening an issue](https://github.com/pandas-dev/pandas/issues/new/choose), which can
be eventually moved to a PDEP.

I think that maybe issues should **always** be opened before submitting a PEP, either as a specific issue or an existing issue concluding that a PEP is required. Maybe we also need to ensure that `@pandas-dev/pandas-core` is also notified in these cases.

If we have some form of gate for opening a PEP, then most should be accepted.

**Review by simonjayhawkins — APPROVED** [review]
Thanks @datapythonista generally lgtm.

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (jorisvandenbossche) [reply web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:57]: > we need a voting process here

I think what Marc proposed above is to defer a discussion about the decision process (basically our governance model) for a subsequent discussion.

(which sounds good to me to keep the discussion manageable)

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (jorisvandenbossche) [reply web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md]: > Regarding merging PDEPs as draft, feels like it could make discussions more difficult.

We will probably have to see how it goes in practice. For many PDEPs a single PR with proposal, discussion, acceptance all in one will quite likely be sufficient. 
But I can imagine for larger/longer-running discussions (I am thinking eg about the copy/view proposal), that it could be useful to already have a hosted version of PDEP that is still under discussion. 

As some references: numpy seems to already merge NEP PRs if they are still under discussion (eg https://github.com/numpy/numpy/pull/18456), but they also have substantial discussion on the mailing list. Similarly in Python, PEPs are merged quickly, but also there the discussion is mostly happening on mailing list or discourse. That's of course different with the current way in pandas where most discussion typically happens on GitHub.

**Review by jorisvandenbossche — COMMENTED** [review]
It might be useful to somewhere list the possible "status" values? (now those seem to be included in the different workflow sections, but it's not very explicit)

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (rhshadrach) [reply web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:89]: This is to contrast rejected PDEPs against accepted PDEPs; i.e. "even rejected PDEPs are useful". I think the wording is okay, but maybe "Rejected PDEPs are just as useful as accepted PDEPs..." would make it more clear?

**Comment by jorisvandenbossche** [comment]
Related to my previous comment (https://github.com/pandas-dev/pandas/pull/47444#discussion_r905868612) about starting a new PDEP as draft / under discussion: this has been changed in the text, but at the same time we still say to create the new PDEP file in the ``web/pdeps/accepted/`` directory. 
Personally, I still find that a bit confusing and sending a potentially wrong message. If we have an `under_discussion` status, why not also reflect that in the directory / path? (although this probably also comes back to whether we have the intention to ever merge PRs in that status, because indeed if we never do that, we would actually never use that directory in practice)

Taking one step back: do we actually need those different directories? (I suppose they are currently used to simplify the script to automate the website content to list the PDEPs) 
For example, NumPy (and also Python PEPs) uses a flat directory for the documents: https://github.com/numpy/numpy/tree/main/doc/neps. And it then infers the status from the header. If we have a standard preamble that has the status in that list, that also sounds doable to automate this (instead of being based on directories)? It would make my above discussion point (about in which directory to start the proposal) moot, and can also simplify the overall process a little bit (eg you never need to move files for changing a status, but just change the Status line in the document (eg when a PDEP gets "implemented")).

File: web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md
Comment (jorisvandenbossche) [reply web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:89]: Ah, I missed the first "as" in the sentence, so I read it as "they are useful as accepted PDEP", instead of "they are _as_ useful as accepted PDEPs", hence the confusion.

**Comment by datapythonista** [comment]
Thanks all for the feedback, very interesting points.

For the voting, I'm working on the governance documents in parallel, and I'll propose changes there to have a way to decide when a PDEP is ready to merge.

For the name to use, there are several comments about names, and many valid points. It's probably not so important, but to find something that most people are happy with, I think we can run a quick poll. I'll write here different options including the ones already proposed. Feel free to suggest more, and when we're ready I'll send a link to the pandas-dev list with a poll.

- PDEP
- pdEP
- PEP
- EP
- PCP (pandas change proposal)
- DEP (dataframe enhancement proposal)

I'll be addressing the rest of the comments in this PR.

**Comment by noatamir** [comment]
May I suggest adding a template for new proposals as part of this PR, or as a follow up if that's preferable? 😇 

I find templates useful to help facilitate structured discussions, and that important points relevant for consideration aren't missed by the proposal author(s) before submitting them, e.g. how would this affect backwards compatibility.  

For reference I include these two templates which could be relevant as a starting point:
- scikit-learn's [SLEP template](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep_template.html)
- NumPy's [NEP template](https://numpy.org/neps/nep-template.html)

**Comment by datapythonista** [comment]
Thanks all for the feedback. Seems like the preferred name is PDEP, so keeping it. I addressed all the comments, there are two pending things, that I think make more sense to address in future PRs:

- How the exact decision on when a PDEP is approved (which is part of a wider discussion about governance)
- Adding a template for PDEPs. While this seems a great idea, I personally think it's premature to have it initially, and I'd wait until we've got few PDEPs, and then decide if we think it adds value, and what should be the format, with concrete examples

Let me know if any other comment. @jreback you reviewed with `Requested changes`, I rewrote that whole section that other people also found misleading, I think your comments should be addressed now.

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (simonjayhawkins) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:88]:
```suggestion
is not part of the roadmap and future plans, but a change that has already
```

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (simonjayhawkins) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:90]: not yet relevant, but would this be just a note in the revision history or something else?

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (simonjayhawkins) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:95]:
```suggestion
not in the best interests of the project. Rejected PDEPs are as useful as accepted
```

File: web/pandas/pdeps/0001-purpose-and-guidelines.md
Comment (simonjayhawkins) [inline web/pandas/pdeps/0001-purpose-and-guidelines.md:108]:
```suggestion
discussion with the author, instead of merging them as rejected. This is to avoid
adding noise to the list of rejected PDEPs, which should contain documentation as
```

**Review by simonjayhawkins — APPROVED** [review]
Thanks @datapythonista

**Review by mroeschke — APPROVED** [review]

File: web/pandas/about/roadmap.md
Comment (jreback) [reply web/pandas/about/roadmap.md:42]: yeah switch to August :->

**Review by jreback — APPROVED** [review]
lgtm. some open comments by @simonjayhawkins

**Review by rhshadrach — APPROVED** [review]
Regarding putting up PDEPs under discussion as open PRs, the one negative aspect is that they aren't protected. The author can force-push making changes hard to follow. I think this will work well assuming no bad actors, and I think that's an okay assumption to make, but wanted to be explicit about it. Also, this downside is outweighed by the upside of being able to have all discussion in a consolidated place in my opinion.

**Review by lithomas1 — APPROVED** [review]

**Comment by datapythonista** [comment]
Thanks all for the reviews and feedback!

**Comment by h-vetinari** [comment]
An unexpected & pleasant surprise that this old issue of mine got fixed. Not sure I have the time/energy to pour those old PDEP-sized ideas into documents, but glad to know that there'd now be a way to do it.

**Comment by simonjayhawkins** [comment]
Thanks @datapythonista for the PR.

> An unexpected & pleasant surprise that this old issue of mine got fixed. Not sure I have the time/energy to pour those old PDEP-sized ideas into documents, but glad to know that there'd now be a way to do it.

Thanks @h-vetinari. feel free to comment on those issues that a PDEP may be a good idea as others may think that they are worth pushing now.

**Comment by simonjayhawkins** [comment]
and if you have already put a lot of time and effort in the discussion, you may have already made it alot easier to progress.

**Pull request merged**
